Mathlib/Algebra/Module/Presentation/Basic.lean

```diff
@@ -0,0 +1,284 @@
+
+/-- The canonical (surjective) linear map `(relations.G →₀ A) →ₗ[A] relations.Quotient`. -/
+def toQuotient : (relations.G →₀ A) →ₗ[A] relations.Quotient := Submodule.mkQ _
+
+@[simp]
+lemma toQuotient_relation (r : relations.R) :
+    relations.toQuotient (relations.relation r) = 0 := by
+  dsimp only [toQuotient]
+  rw [Submodule.mkQ_apply, Submodule.Quotient.mk_eq_zero]
+  exact Submodule.subset_span (by simp)
+
+/-- The linear map `(relations.R →₀ A) →ₗ[A] (relations.G →₀ A)` corresponding to the relations
+given by `relations : Relations A`. -/
+noncomputable def map : (relations.R →₀ A) →ₗ[A] (relations.G →₀ A) :=
+  Finsupp.lift _ _ _ relations.relation
+
+@[simp]
+lemma map_single (r : relations.R) :
+    relations.map (Finsupp.single r 1) = relations.relation r := by
+  simp [map]
+
+variable (M : Type v) [AddCommGroup M] [Module A M]
+
+/-- The type of solutions in a module `M` of the equations given by `relations : Relations A`. -/
+@[ext]
+structure Solution where
+  /-- the image in `M` of each variable -/
+  var (g : relations.G) : M
+  relation (r : relations.R) : Finsupp.lift _ A _ var (relations.relation r) = 0
+
+namespace Solution
+
+variable {relations M}
+
+section
+
+variable (solution : relations.Solution M)
+
+/-- Given `relations : Relations A` and a solution in `relations.Solution M`, this is
+the linear map `(relations.G →₀ A) →ₗ[A] M` canonically associated to the . -/
+noncomputable def π : (relations.G →₀ A) →ₗ[A] M := Finsupp.lift _ _ _ solution.var
+
+@[simp]
+lemma π_single (g : relations.G) :
+    solution.π (Finsupp.single g 1) = solution.var g := by simp [π]
+
+@[simp]
+lemma π_relation (r : relations.R) : solution.π (relations.relation r) = 0 := solution.relation r
+
+@[simp]
+lemma π_comp_map : solution.π.comp relations.map = 0 := by aesop
+
+@[simp]
+lemma π_comp_map_apply (x : relations.R →₀ A) : solution.π (relations.map x) = 0 := by
+  change solution.π.comp relations.map x = 0
+  rw [π_comp_map, LinearMap.zero_apply]
+
+/-- Given `relations : Relations A` and `solution : relations.Solution M`,
+this is the canonical linear map from `relations.R →₀ A` to the kernel
+of `solution.π : (relations.G →₀ A) →ₗ[A] M`. -/
+noncomputable def mapToKer : (relations.R →₀ A) →ₗ[A] (LinearMap.ker solution.π) :=
+  LinearMap.codRestrict _ relations.map (by simp)
+
+@[simp]
+lemma mapToKer_coe (x : relations.R →₀ A) : (solution.mapToKer x).1 = relations.map x := rfl
+
+lemma span_relation_le_ker_π :
+    Submodule.span A (Set.range relations.relation) ≤ LinearMap.ker solution.π := by
+  rw [Submodule.span_le]
+  rintro _ ⟨r, rfl⟩
+  simp only [SetLike.mem_coe, LinearMap.mem_ker, π_relation]
+
+/-- Given `relations : Relations A` and `solution : relations.Solution M`, this is
+the canonical linear map `relations.Quotient →ₗ[A] M` from the module. -/
+noncomputable def fromQuotient : relations.Quotient →ₗ[A] M :=
+  Submodule.liftQ _ solution.π solution.span_relation_le_ker_π
+
```

Comment:
Should there be a (simp) lemma `solution.fromQuotient (Submodule.Quotient.mk x) = solution.π x`?

Reply:
Yes. Indeed, it makes unnecessary the lemma:
```lean
lemma fromQuotient_single (g : relations.G) :
    solution.fromQuotient (Submodule.Quotient.mk (Finsupp.single g 1)) =
      solution.var g := by
  simp
```